lib/iris/fileformats/grib/message.py

```diff
@@ -52,31 +66,53 @@ def messages_from_filename(filename):
             Name of the file to generate fields from.
 
         """
-        with open(filename, 'rb') as grib_fh:
-            while True:
-                offset = grib_fh.tell()
-                grib_id = gribapi.grib_new_from_file(grib_fh)
-                if grib_id is None:
-                    break
-                raw_message = _RawGribMessage(grib_id)
-                recreate_raw = _MessageLocation(filename, offset)
-                yield _GribMessage(raw_message, recreate_raw)
-
-    def __init__(self, raw_message, recreate_raw):
+        grib_fh = open(filename, 'rb')
+        # create an _OpenFileRef to manage the closure of the file handle
+        file_ref = _OpenFileRef(grib_fh)
+
+        while True:
+            offset = grib_fh.tell()
+            grib_id = gribapi.grib_new_from_file(grib_fh)
+            if grib_id is None:
+                break
+            raw_message = _RawGribMessage(grib_id)
+            recreate_raw = _MessageLocation(filename, offset)
+            yield GribMessage(raw_message, recreate_raw, file_ref=file_ref)
+
+    def __init__(self, raw_message, recreate_raw, file_ref=None):
         """
-
-        Args:
-
-        * raw_message:
-            The _RawGribMessage instance which should be wrapped to
-            provide the `data` attribute.
+        It is recommended to obtain GribMessage instance from the static method
+        :meth:`GribMessage.messages_from_filename`, rather than creating
+        them directly.
 
         """
+        # A RawGribMessage giving gribapi access to the original grib message.
         self._raw_message = raw_message
+        # A _MessageLocation which biggus uses to read the message data array,
+        # by which time this message may be dead and the original grib file
+        # closed.
         self._recreate_raw = recreate_raw
+        # An _OpenFileRef to keep the grib file open while this GribMessage is
+        # alive, so that we can always use self._raw_message to fetch keys.
+        self._file_ref = file_ref
 
     @property
     def sections(self):
+        """
+        Return the key-value pairs of the message keys, grouped by containing
+        section.
+
+        Key-value pairs are collected into a dictionary of
```

Comment:
Maybe doesn't read that clearly. 
"collected into" is describing from the implementation POV rather than usage.
Maybe something like this?..
```
Sections in a message are indexed by GRIB section-number, and values in a section are indexed by key strings.

.. For example::

    print(grib_message.sections[4]['parameterNumber'])
    grib_message.sections[1]['minute'] = 0
```